daphne_worker/src/durable/reports_processed.rs

```diff
@@ -53,18 +54,66 @@ pub struct ReportsProcessed {
     alarmed: bool,
 }
 
-impl ReportsProcessed {
-    /// Check if the report has been processed. If not, return None; otherwise, return the ID.
-    async fn to_checked(&self, report_id: ReportId) -> Result<Option<ReportId>> {
-        let key = format!("processed/{}", report_id.to_hex());
-        let processed: bool = state_set_if_not_exists(&self.state, &key, &true)
-            .await?
-            .unwrap_or(false);
-        if processed {
-            Ok(Some(report_id))
-        } else {
-            Ok(None)
+#[derive(Debug, Clone)]
+struct ReportIdKey<'s>(&'s ReportId, String);
+fn report_id_to_do_key(id: &ReportId) -> ReportIdKey<'_> {
+    ReportIdKey(id, format!("processed/{}", id.to_hex()))
+}
+
+#[derive(Debug)]
+enum CheckedReplays<'s> {
+    Replayed(Vec<&'s ReportId>),
+    NotReplayed(Vec<ReportIdKey<'s>>),
+}
+
+impl<'r> Default for CheckedReplays<'r> {
+    fn default() -> Self {
+        Self::NotReplayed(vec![])
+    }
+}
+
+impl<'r> CheckedReplays<'r> {
+    pub fn add_replay(mut self, id: &'r ReportId) -> Self {
+        match &mut self {
+            Self::Replayed(r) => {
+                r.push(id);
+                self
+            }
+            Self::NotReplayed(_) => Self::Replayed(vec![id]),
+        }
+    }
+
+    pub fn add_not_replayed(mut self, id: ReportIdKey<'r>) -> Self {
+        match &mut self {
+            Self::Replayed(_) => {}
+            Self::NotReplayed(r) => r.push(id),
         }
+        self
+    }
+}
+
+impl ReportsProcessed {
+    async fn check_replays<'s>(&self, report_ids: &'s [ReportId]) -> Result<CheckedReplays<'s>> {
+        futures::stream::iter(report_ids.iter().map(report_id_to_do_key))
+            .then(|id| {
+                let state = &self.state;
+                async move {
+                    state_get::<bool>(state, &id.1)
+                        .await
+                        .map(|presence| match presence {
+                            // if it's present then it's a replay
+                            Some(_) => Err(id.0),
```

Comment:
Suggestion: I'd be a bit stricter here. Maybe handle `Some(false)` the same way as `None`?
```suggestion
                            Some(true) => Err(id.0),
```

Reply:
Nice idea.